zeroconf/__init__.py

```diff
@@ -1203,41 +1203,59 @@ def __init__(self, zc: 'Zeroconf') -> None:
         self.readers = {}  # type: Dict[socket.socket, Listener]
         self.timeout = 5
         self.condition = threading.Condition()
+        self.socketpair = socket.socketpair()
         self.start()
 
     def run(self) -> None:
         while not self.zc.done:
             with self.condition:
-                rs = self.readers.keys()
+                rs = list(self.readers.keys())
                 if len(rs) == 0:
                     # No sockets to manage, but we wait for the timeout
                     # or addition of a socket
                     self.condition.wait(self.timeout)
 
             if len(rs) != 0:
                 try:
+                    rs = rs + [self.socketpair[0]]
                     rr, wr, er = select.select(cast(Sequence[Any], rs), [], [], self.timeout)
                     if not self.zc.done:
                         for socket_ in rr:
                             reader = self.readers.get(socket_)
                             if reader:
                                 reader.handle_read(socket_)
 
+                        if self.socketpair[0] in rr:
+                            # Clear the socket's buffer
+                            self.socketpair[0].recv(128)
+
                 except (select.error, socket.error) as e:
                     # If the socket was closed by another thread, during
                     # shutdown, ignore it and exit
                     if e.args[0] not in (errno.EBADF, errno.ENOTCONN) or not self.zc.done:
                         raise
+        try:
+            self.socketpair[0].close()
+            self.socketpair[1].close()
+        except socket.error:
+            pass
+
+    def _notify(self) -> None:
+        self.condition.notify()
+        try:
+            self.socketpair[1].send(b'x')
+        except socket.error:
+            pass
```

Comment:
In what scenarios will those `send()` and `close()` calls actually throw exceptions?

Reply:
`send()` may close during shutdown, here's an example: https://travis-ci.org/github/jstasiak/python-zeroconf/jobs/674340270.

the `close()` should not throw, and if it does I guess it won't matter because it happens when the worker thread is done.

I cleaned it up a bit accordingly.